### ISSUE TITLE
Support for asm-sources, cmm-sources, and js-sources in executable components

### ISSUE THREAD

**Issue by hsyl20** [haskell/cabal#8639]
**Describe the bug**

Cabal doesn't support `asm-sources`, `cmm-sources`, and `js-sources` in `executable` components.

**To Reproduce**

```bash
$ mkdir cabal-foreign
$ cd cabal-foreign
$ cabal init --cabal-version=3.0
$ cabal build ; works fine
$ echo "    asm-sources: whatever.s" >> cabal-foreign.cabal
$ echo "    js-sources: whatever.js" >> cabal-foreign.cabal
$ echo "    cmm-sources: whatever.cmm" >> cabal-foreign.cabal
$ cabal clean && cabal build ; should fail because of the missing files, but doesn't
```

**Expected behavior**

Cabal should take into account the specified foreign files: link them when present, or fail when not present as in the reproducer above.

At the very least a warning that the fields are parsed but unused would be a better UX.

**System information**
 - Linux
 - cabal-install 3.6.2.0 and HEAD
 - ghc 9.2.4

**Additional context**

I've noticed this bug while I was adapting the implementation of `asm-sources` to fix the support for `js-sources` in https://github.com/haskell/cabal/pull/8636.

As far as I can tell, the issue is in `Distribution.Simple.GHC` module. `gbuildSources` only returns C and C++ foreign sources (cf fields of `BuildSources` record). It should also return `Cmm`, `asm`, and `js` sources. Then `BuildSources` new fields must be handled properly in `gbuild`.

It looks like the code to build C/C++ sources for a library (in `buildOrReplLib`) is duplicated in `gbuild` to build an executable. I was reluctant to copy-paste the code handling asm/cmm/js sources from `buildOrReplLib` to `gbuild` too. It should probably be factorized instead. That's why I didn't do it as part of https://github.com/haskell/cabal/pull/8636.

**Comment by chreekat** [comment]
I just ran into this :D I'll see what I can slap together during Zurihac.

**Comment by mpickering** [comment]
It's pretty unclear why `buildAndReplLib` and `gbuild` are separate functions (for reasons like this).

**Comment by ysangkok** [comment]
can this be closed now since #9061 was merged?

**Comment by hsyl20** [comment]
Indeed!
